std/range/package.d

```diff
@@ -5369,284 +5369,387 @@ pure @safe nothrow unittest
     ---
 */
 auto iota(B, E, S)(B begin, E end, S step)
-if ((isIntegral!(CommonType!(B, E)) || isPointer!(CommonType!(B, E)))
-        && isIntegral!S)
 {
-    import std.conv : unsigned;
+    static if (!is(B == Unqual!B) || !is(E == Unqual!E) || !is(S == Unqual!S))
+    {
+        return iota(Unqual!B(begin), Unqual!E(end), Unqual!S(step));
+    }
+    else
+    {
+        return IotaResult!(B, E, S)(begin, end, step);
+    }
+}
 
-    alias Value = CommonType!(Unqual!B, Unqual!E);
-    alias StepType = Unqual!S;
+/// Ditto
+auto iota(B, E)(B begin, E end)
+{
+    static if (!is(B == Unqual!B) || !is(E == Unqual!E))
+    {
+        return iota(Unqual!B(begin), Unqual!E(end));
+    }
+    else static if (is(typeof(begin + B(1))))
+    {
+        return iota(begin, end, B(1));
+    }
+    else static if (is(typeof(begin + 1)))
+    {
+        return iota(begin, end, 1);
+    }
+    else static if (is(typeof(++begin)))
```

Comment:
this should go before the test `begin + 1` i.e. prefer increment to adding 1

Reply:
Wait... I'm not sure you're right about that. A more correct test would put `begin + 1` first, and check in `popFront()` if it can use `begin++`. That way we can have efficient slicing on types that support both.